### ISSUE TITLE
Literal math with repeated variables can lead to false negatives

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#2870]
**Describe the bug**
Literal math doesn't handle repeated variables correctly.

**To Reproduce**
```python
from typing import Literal


def func(x: Literal[2, 3]):
    a = x * x
    reveal_type(a)  # Literal[4, 6, 9]
    if a == 6:
        six: Literal[6] = a  # no error; false negative
 ```

**Expected behavior**
`a` is `Literal[4, 9]`

**VS Code extension or command-line**
pyright 1.1.209 linux commandline

**Additional context**
This is not entirely unlike this kind of case, which pyright does not support:
```python
def func(x: int | str):
    x + x
```
However, this might be a bit easier for the case of literal math.

**Comment by JelleZijlstra** [comment]
Pyright's behavior in your code sample looks right to me. Did you mean to use `!=`?

**Comment by GBeauregard** [comment]
> Pyright's behavior in your code sample looks right to me. Did you mean to use `!=`?

oh, I was just trying to clear up the other cases to create a false negative but it's unreachable anyway, the main problem is the inclusion of a 6

**Comment by erictraut** [comment]
From a math perspective, I can see why you'd think that these results are incorrect. From a type perspective, the results are correct.
